[BD-29] [TNL-7209] [BB-2595] Enable Anonymous access for course blocks API

xitij2000 opened the pull request:
This PR adds support for accessing the course blocks for a public course as an anonymous user. 

To access course blocks for a public course anonymously, you simply need to pass an empty value for the `username` parameter. e.g.
- ``/api/courses/v1/blocks/?course_id=<course_id>&depth=all&username=``
- ``/api/courses/v1/blocks/?course_id=<course_id>&depth=all&username``

**JIRA tickets**: [TNL-7209](https://openedx.atlassian.net/browse/TNL-7209)

**Sandbox URL**: 

* LMS: https://pr24264.sandbox.opencraft.hosting/
* Studio: https://studio.pr24264.sandbox.opencraft.hosting/


**Merge deadline**: None

**Testing instructions**:

1. Create a public course or make an existing course public
2. In a private/anonymous session, access the course blocks API for this course with an empty username parameter:
   `/api/courses/v1/blocks/?course_id=<course_id>&depth=all&username=`
   Note: It's important to provide the username parameter, it just needs to be empty.
3. In a session logged in as a user not enrolled in this course access the same url as above.
4. In both cases you should be able to see the course blocks. 

Testing on sandbox:
- Public course:
https://pr24264.sandbox.opencraft.hosting/api/courses/v1/blocks/?course_id=course-v1%3AedX%2BDemoX%2BPublic_Demo_Course&depth=all&username

| User type  | Public Course                                                                                                                 | Regular Course                                                                                                       |
|------------|-------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------|
| Anonymous  | all_blocks=True not allowed (403). Only empty username is allowed. All other values result in a 403.                          | No access (403)                                                                                                      |
| Unenrolled | all_blocks=True not allowed (403). Only empty username is allowed. All other values (including own username) result in a 403. | No access (403)                                                                                                      |
| Enrolled   | all_blocks=True not allowed (403). Only empty username or own username is allowed. All other values result in a 403.          | all_blocks=True not allowed (403). Only empty username or own username is allowed. All other values result in a 403. |
| Staff       | all_blocks=True is allowed. Takes precedence over username. All usernames permitted.                                          | all_blocks=True is allowed. Takes precedence over username. All usernames permitted.                                 |

**Author Notes**:

- Does it perhaps make sense to cache the course visibility in the course overview?


**Reviewers**
- [ ] @lgp171188 
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
```

edx-webhook commented:
Thanks for the pull request, @xitij2000! I've created [OSPR-4735](https://openedx.atlassian.net/browse/OSPR-4735) to keep track of it in JIRA, where we prioritize reviews.  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Please let us know once your PR is ready for our review and all tests are green.

natabene commented:
@xitij2000 Thank you for your contribution. Please let me know once it is ready for our review.

lgp171188 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,20 +108,51 @@ def clean(self):
+        if cleaned_data.get('all_block', None) is not None:
+            raise PermissionDenied(
+                u"Anonymous users do not have permission to access all blocks in '{course_key}'."
+                    .format(course_key=six.text_type(course_key))
```
@xitij2000, can you fix the code quality errors reported on this line and 2 other occurrences of the same issue in the indentation of the line containing `.format`?

lgp171188 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,20 +108,51 @@ def clean(self):
+        if not is_course_public(get_course_by_id(course_key, depth=0)):
+            raise PermissionDenied(
+                u"Course blocks for '{course_key}' cannot be accessed anonymously."
+                    .format(course_key=course_key)
```
Same as https://github.com/edx/edx-platform/pull/24264/files#r443771422

lgp171188 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -158,11 +191,16 @@ def _verify_other_user(requesting_user, requested_username, course_key):
             raise PermissionDenied(
                 u"'{requesting_username}' does not have permission to access view for '{requested_username}'."
-                .format(requesting_username=requesting_user.username, requested_username=requested_username)
+                    .format(requesting_username=requesting_user.username, requested_username=requested_username)
```
Same as https://github.com/edx/edx-platform/pull/24264/files#r443771422

lgp171188 commented on lms/djangoapps/course_api/blocks/permissions.py:
```diff
@@ -30,5 +30,6 @@ def can_access_self_blocks(requesting_user, course_key):
         (requesting_user.id and CourseEnrollment.is_enrolled(requesting_user, course_key)) or
-        has_access(requesting_user, CourseStaffRole.ROLE, course_key)
+        has_access(requesting_user, CourseStaffRole.ROLE, course_key) or
+        is_course_public(get_course_by_id(course_key, depth=0))
```
@xitij2000, `get_course_by_id()` will raise a `Http404` exception and that is causing [this unit test failure](https://build.testeng.edx.org/job/edx-platform-python-pipeline-pr/18734/testReport/junit/lms.djangoapps.course_api.blocks.tests.test_forms/TestBlockListGetForm/Run_Tests___lms_unit___test_non_existent_usage_key/). Should we wrap that call in a `try-except` call and return a boolean based on what happens?

lgp171188 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,20 +108,51 @@ def clean(self):
+        if cleaned_data.get('username', None) is not '':
+            raise PermissionDenied(u"Anonymous users cannot access another user's blocks.")
+
+        if not is_course_public(get_course_by_id(course_key, depth=0)):
```
@xitij2000, similar to my comment on the usage of `get_course_by_id()` in a review comment below, the `Http404` exception raised by `get_course_by_id()` may have to be handled if we don't want that to propagate.

lgp171188 replied:
@xitij2000, just wanted to confirm if it is okay to let the `Http404` exception to propagate in this case.

lgp171188 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,20 +108,51 @@ def clean(self):
+        """
+        Verifies form for when the requesting user is anonymous.
+        """
+        if cleaned_data.get('all_block', None) is not None:
```
@xitij2000, the field is called `all_blocks`.

lgp171188 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,20 +108,51 @@ def clean(self):
+                    .format(course_key=six.text_type(course_key))
+            )
+
+        if cleaned_data.get('username', None) is not '':
```
@xitij2000, when I send a request as an anonymous user to http://localhost:18000/api/courses/v2/blocks/?course_id=course-v1%3AedX%2BDemoX%2BDemo_Course, this condition is true and the below error message is returned. However, I haven't passed either the `username` field or the `all_blocks` field and in that case, the API should return the error about requiring one of those two fields to retain backward compatibility.

lgp171188 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,20 +107,51 @@ def clean(self):
 
+    def clean_username(self):
+        if 'username' in self.data:
+            return self.cleaned_data['username']
```
@xitij2000, a request to http://localhost:18000/api/courses/v2/blocks/?course_id=course-v1%3AedX%2BDemoX%2BDemo_Course&depth=all&username= as a non-staff user in the current `master` branch returns a `"This field is required unless all_blocks is requested."` error. However with this change, the API returns `"'audit' does not have permission to access view for ''"`

xitij2000 replied:
This is by design. This is the new behaviour. It is not backwards-compatible to this extent. While I could add a special case and have it say <anonymous user> instead of a blank string, I think the current behaviour is also OK, so there is no need for additional complexity. 

While earlier a "username=" was considered an invalid/blank input, it is no longer considered that, so the old error is no longer valid.

lgp171188 replied:
@xitij2000, I don't have the permission to mark this as resolved. Can you do that?

lgp171188 commented:
@xitij2000, I  tested the changes in the following scenarios (based on my assumption of the expected behaviour) and below are the results. Please let me know if the expected behaviour for the items with ❌ or :question: are incorrect. If all of these are correct, can you fix the issues marked with ❌?

* Course with public access enabled
  * Anonymous user:
  :heavy_check_mark:  Can access the course blocks by passing `username=`
  ❌ Cannot specify a `username` :heavy_check_mark:  or `all_blocks` ❌ (can specify it when specifying an empty username for anonymous access).
  * Authenticated non-staff user:
:heavy_check_mark: Can access the course blocks by passing `username=` or `username=<their username>`
❌ Cannot specify other usernames for the `username` parameter :heavy_check_mark: or `all_blocks` ❌ (can specify it when specifying an empty username like an anonymous user would).
  * Authenticated staff user:
:heavy_check_mark: Can access the course blocks by passing `username=` or `username=<their username>`
:heavy_check_mark:  Cannot specify other usernames for the `username` parameter or `all_blocks` 

* Course without public access enabled or just public outline access enabled
  * Anonymous user:
  :heavy_check_mark:  Cannot access the course blocks by passing `username=`
  :heavy_check_mark: Cannot specify a `username` or `all_blocks`.
  * Authenticated non-staff user:
:heavy_check_mark: Can access the course blocks by passing `username=` or `username=<their username>`
:heavy_check_mark: Cannot specify other usernames for the `username` parameter or `all_blocks`.
:question: Cannot specify an empty username (returns a `"'audit' does not have permission to access view for ''."` error when using the `audit` user account).
  * Authenticated staff user:
:heavy_check_mark: Can access the course blocks by passing `username=` or `username=<their username>`
:heavy_check_mark: Cannot specify other usernames for the `username` parameter or `all_blocks`.
:question: Cannot specify an empty username (returns a `Requested user '' does not exist.` error).

lgp171188 commented:
@xitij2000, thanks for the detailed explanation. I have some follow-up questions.

> all_blocks and username are mutually exclusive.

The two items marked in my previous comment with a ❌ exhibit a behaviour, where it is possible to pass `username=''` and `all_blocks=1` at the same time and it works without any errors. Doesn't this contradict the expectation that the `username` and `all_blocks` field are mutually exclusive?

lgp171188 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,30 +107,63 @@ def clean(self):
         return cleaned_data
 
+    def clean_username(self):
+        if 'username' in self.data:
```
@xitij2000, can you add a docstring for this method?

stvstnfrd commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -138,7 +179,6 @@ def _verify_no_user(requesting_user, cleaned_data, course_key):
             )
 
         # return None for user
-        return None
```
Why remove the `return` but not the comment?

xitij2000 replied:
I guess a return None is implicit. I don't quite remember the reason, perhaps it was due to a linting issue? 
Will add it back and see.

xitij2000 replied:
@stvstnfrd I'm specifically talking about the platform's linting rules. I completely agree that an explicit None is beetter in many cases. 

I can add it back with an exception?

stvstnfrd commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -158,6 +198,11 @@ def _verify_other_user(requesting_user, requested_username, course_key):
         """
+        # If accessing a public course, and requesting only content available publicly,
+        # we can allow the request.
+        if is_course_public(get_course_by_id(course_key, depth=0)) and requested_username == '':
```
If we swap the order in this comparison, we'll save ourselves the `is_course_public` lookup if the user isn't anonymous (`requested_username != ''`).

```suggestion
        if requested_username == '' and is_course_public(course_key):
```

stvstnfrd commented on lms/djangoapps/course_api/blocks/permissions.py:
```diff
@@ -28,7 +28,14 @@ def can_access_self_blocks(requesting_user, course_key):
+        return user_is_enrolled_or_staff
+    else:
+        try:
+            return is_course_public(get_course(course_key, depth=0))
```
```suggestion
            return is_course_public(course_key)
```

stvstnfrd commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,30 +107,70 @@ def clean(self):
+        if not username == '':
+            raise PermissionDenied(u"Anonymous users cannot access another user's blocks.")
+
+        if not is_course_public(get_course_by_id(course_key, depth=0)):
```
```suggestion
        if not is_course_public(course_key):
```

stvstnfrd commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,30 +107,70 @@ def clean(self):
+        """
+        if 'username' in self.data:
+            return self.cleaned_data['username']
+        else:
```
If an `if` block ends with a `return` statement, the `else` is superfluous and we can remove the `else`, unindent its body, and make it a little easier to read.

stvstnfrd commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,30 +107,70 @@ def clean(self):
+                .format(course_key=six.text_type(course_key))
+            )
+
+        if not username == '':
```
This is "un-pythonic" (non-idiomatic Python)

```suggestion
        if username != '':
```

stvstnfrd commented on lms/djangoapps/course_api/blocks/permissions.py:
```diff
@@ -28,7 +28,14 @@ def can_access_self_blocks(requesting_user, course_key):
     )
+    if user_is_enrolled_or_staff:
+        return user_is_enrolled_or_staff
+    else:
```
Similar to the earlier comment, this one will be even easier to read w/out the `else` (keyword), since it has further indent levels.

stvstnfrd commented on lms/djangoapps/course_api/blocks/permissions.py:
```diff
@@ -28,7 +28,14 @@ def can_access_self_blocks(requesting_user, course_key):
+    else:
+        try:
+            return is_course_public(get_course(course_key, depth=0))
+        except ValueError:
```
Do we need to worry about catching this anywhere else we call this?

xitij2000 replied:
ValueError is raised by get_course if the course is not found. I don't think this should be caught at the `is_course_public` level so this seems to be the best place. 
I don't see an issue in this case because the main scenario for raising the error here is a missing course and in that case `ACCESS_DENIED` seems to make most sense.

stvstnfrd replied:
Sounds good, just wanted to flag it, as it stood out.

xitij2000 commented:
@stvstnfrd Thanks for the review! I've updated the PR based on your suggestions.

Does it make sense to add this use case to the API documentation of the course blocks API view?

xitij2000 commented on lms/djangoapps/course_api/blocks/permissions.py:
```diff
@@ -24,11 +29,25 @@ def can_access_others_blocks(requesting_user, course_key):
+        return ACCESS_DENIED
+
+
+def is_course_public(course_key: CourseKey) -> AccessResponse:
```
Moved this here, because importing get_course in access_utils leads to a whole range of circular imports. If it really should be in access_utils I can make an import in this function to avoid the circular imports.

stvstnfrd commented on lms/djangoapps/course_api/blocks/tests/test_views.py:
```diff
@@ -148,14 +150,73 @@ def assert_true_iff(self, expression, predicate):
+        Verify behaviour when accessing course blocks of a public course for another user anonymously.
+        """
         self.client.logout()
-        self.verify_response(401)
```
It looks like we will no longer have any test cases here that check for a `401 Unauthorized` response;
are there any use cases where this could/should arise?

If so, can we add a relevant test case?

xitij2000 replied:
I can't think of any situation where it would rise up. Now that this API is supported for unauthenticated users, this error isn't being raised explicitly, and I'm not aware of any explicit situations. 

I guess certain requests that are currently 403 Unauthorized, could alternatively be 401 Unauthenticated when raised in case of an anonymous user. 

I'm also wondering if it makes sense to throttle this API now that it's available publicly?

stvstnfrd replied:
@ormsbee Thoughts on throttling? That's a little out of my wheelhouse, presently.

stvstnfrd commented on lms/djangoapps/courseware/access_utils.py:
```diff
@@ -8,17 +8,15 @@
 )
 from lms.djangoapps.courseware.masquerade import get_course_masquerade, is_masquerading_as_student
-from openedx.core.djangoapps.util.user_messages import PageLevelMessages
-from openedx.core.djangolib.markup import HTML
```
All changes in this file seem unrelated?

Do you have an auto-linter enabled in your text editor?

xitij2000 replied:
Sorry there were changes in this file, but they were refactored out. I think is_course_public was here. Will remove.

stvstnfrd commented:
@xitij2000 Would you mind adding a little table/summary/write-up/something that breaks down how this is expected to work in the various configurations?

I know some of this is included earlier in-thread, but it's hard for me to see the final product, given the changes and back-and-forth.

Thanks!

xitij2000 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,40 +105,76 @@ def clean(self):
+        allows us to differentiate between no username being provided (None) vs
+        an empty username being provided ('').
+        """
+        if 'username' in self.data:
```
The reason for that is that we want to be able to distinguish between username = '' and username not being provided at all. The way Django forms code works is that if username is simply not provided as a query parameter, it will show up as an empty string in the cleaned_data, which is what we generally use. So we are unable to distinguish between the case where username is an empty string and when username is absent.

So what I"m doing here is:
- if username is absent, set username to None instead of an empty string
- If username is present use whatever the default string cleaning code returns.
If a

xitij2000 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,40 +105,76 @@ def clean(self):
 
-        if not requested_username:
-            return self._verify_no_user(requesting_user, cleaned_data, course_key)
+        if requested_username is not None and all_blocks:
```
Sure, I think that makes sense, it will also simplify the code.

xitij2000 commented on lms/djangoapps/course_api/blocks/forms.py:
```diff
@@ -106,40 +105,76 @@ def clean(self):
+                .format(course_key=six.text_type(course_key))
+            )
 
+        if username != '':
```
I think the idea is that for anonymous users, the only supported value of username is an empty string. Not providing a username (username=None) is invalid unless all_blocks is specified, specifying that is not allowed for non-staff, any other username is also not allowed for non-staff. 

Honestly, I think some other constant might allow simplifying the code, for instance "@anonymous@" or some other string that's usually invalid in usernames.

xitij2000 commented on lms/djangoapps/course_api/blocks/permissions.py:
```diff
@@ -24,11 +29,25 @@ def can_access_others_blocks(requesting_user, course_key):
+        return ACCESS_GRANTED
+    try:
+        return is_course_public(course_key)
+    except ValueError:
```
is_course_public in calls get_course which raises this error:
https://github.com/edx-solutions/edx-platform/blob/6077359a07972d8ec81f1a6ca2e414bfec79e71a/lms/djangoapps/courseware/courses.py#L52-L65

xitij2000 commented:
> Would you mind adding a little table/summary/write-up/something that breaks down how this is expected to work in the various configurations?

Sure, I've added this. There seem to be some issues with the sandbox, will try to get it working.

xitij2000 commented:
jenkins run python
jenkins run py38 python

xitij2000 commented:
@stvstnfrd I've updated the PR to include your suggestions. I've also fixed the sandbox and added a [public course ther](https://pr24264.sandbox.opencraft.hosting/courses/course-v1:edX+DemoX+Public_Demo_Course/)

stvstnfrd reviewed (approved):
Looks good, thanks @xitij2000 ! I tagged @ormsbee to hear if he has thoughts on throttling, but otherwise, I'm +1.

ormsbee commented:
@stvstnfrd: Caching thoughts:

1. This is going to be an expensive operation, and it is very possible that it could get crushed in its current form before we run up to the middleware or CDN-enforced limits.
2. But anonymous access at this endpoint is also massively cachable. It might be as straightforward as making sure we send the right headers when running in this mode, so that it can be cached at the CDN layer. I'm not sure if there's anything we'd have to toggle at the Cloudflare layer to make sure it's querystring-aware and doesn't start caching user-specific stuff.

xitij2000 commented:
@ormsbee Does it make sense to cache the course visibility in the overview? The only reason we're touching the modulestore in this API is to get the course visibility.

ormsbee commented:
@xitij2000: I think it makes sense to add it to the CourseOverview. I keep forgetting if there's anything that needs to be done to roll out updates to that model aside from adding the field, making the migration, populating on the publish handler, and incrementing the version on the class. In any case, it's desirable to add it to a data model irrespective of performance, since it's not currently queryable in any sane way on aggregates, and CourseOverview is the most convenient target for that at the moment.

That being said, running Block Structures is itself an extremely expensive operation, so from the performance point of view, nothing is going to beat top level caching of the anonymous output. So if you have to prioritize one or the other, please do the caching.

xitij2000 commented:
@ormsbee There seems to be a mechanism to do this using Django's cache with [this utility](https://github.com/edx/edx-platform/blob/master/common/djangoapps/util/cache.py#L25). I'm investigating how to do it using cache headers, if it's possibel to do that way.

marcotuts commented:
@xitij2000 - I wanted to check in here, this (and possibly 1 other ticket) are all that is left in BD-29 so I wanted to see if there's anything you needed from us. (This might not be scheduled in your current sprint and if so that's fine, just wanted to check.)

xitij2000 commented:
> @xitij2000 - I wanted to check in here, this (and possibly 1 other ticket) are all that is left in BD-29 so I wanted to see if there's anything you needed from us. (This might not be scheduled in your current sprint and if so that's fine, just wanted to check.)

This is in my current sprint, I was just investigating the best approach here. I think I might have a solution. 

@ormsbee I think I was overthinking this a bit by trying to come up with a way to get this cached by the CDN in a way that it won't also affect live users. i.e. have the CDN layer cache the response for an anonymous user and serve that to a live user. What I've done instead is to just add the cache header in case the username parameter is a blank string (i.e. an anonymous user's data is requested). This will enable caching for anonymous and authenticated users, but the response should be the same, so I don't think that should be a concern? 

Does this approach seem OK? Anonymous users requesting course data for non-public courses will still call the is_public check but that results in a 403 and AFAIK that shouldn't be cached, so I think adding the public/private status of a course to the outline could still help there.

ormsbee commented:
> @ormsbee I think I was overthinking this a bit by trying to come up with a way to get this cached by the CDN in a way that it won't also affect live users. i.e. have the CDN layer cache the response for an anonymous user and serve that to a live user. What I've done instead is to just add the cache header in case the username parameter is a blank string (i.e. an anonymous user's data is requested). This will enable caching for anonymous and authenticated users, but the response should be the same, so I don't think that should be a concern?
> 
> Does this approach seem OK? Anonymous users requesting course data for non-public courses will still call the is_public check but that results in a 403 and AFAIK that shouldn't be cached, so I think adding the public/private status of a course to the outline could still help there.

Yes, that approach sounds fine. My only concern was whether our caching configuration [respected query strings](https://support.cloudflare.com/hc/en-us/articles/200168256-Understand-Cloudflare-Caching-Level), but I've just checked and we use "Standard", so we should be fine there.

xitij2000 commented:
@stvstnfrd I think the caching scenario is covered as well now. Do tell if I should add the course visibility to the outline data. I don't think I will be able to do that this sprint though, so if needed I can work on that in a separate PR in the next sprint?

ormsbee commented on lms/djangoapps/course_api/blocks/views.py:
```diff
@@ -226,11 +227,17 @@ def list(self, request, usage_key_string, hide_access_denials=False):  # pylint:
+            # data about the anonymous view of a course, which can be cached. In this
+            # case we add the usual caching headers to the response.
+            if params.cleaned_data.get('username', None) == '':
+                patch_response_headers(response)
```
For my paranoia, can you also explicitly check to make sure we don't cache when `all_blocks=True`, even if the caller also does `username=` in the URL?

xitij2000 replied:
@ormsbee  When `all_blocks=True` the code explicitly sets username to None, so this code would not trigger in that case.

If you use `all_blocks` as anything other than a staff user you will get a 403. If you use it along with `username`, the username will be set to None, so this code will never trigger. 

Relevant code: https://github.com/edx/edx-platform/pull/24264/files#diff-fc18b4a3c351e9d906f21cc6def6610eR117-R118

That said, to avoid any future I've added a test to specifically test for the situation where both `username=''` and `all_blocks=True`. It tests the response as a regular enrolled user, an unenrolled user, an anonymous user and staff user. In the first three it asserts that the response is 403 and not cached, and for staff, that the response is correct and not cached.

stvstnfrd commented:
@xitij2000 Looks good to me! Want to squash that last commit in and then we can merge?

xitij2000 commented:
@stvstnfrd I've squashed the commits down to one. I also updated the documentation to include the fact that an empty/blank username field is now valid.

xitij2000 commented:
jenkins run py38 python
jenkins run py38 quality
